spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaConsumerBackoffManager.java

```diff
@@ -47,87 +53,162 @@
 		removeBackoff(context.topicPartition);
 	}
 
-	private boolean isNotDue(long dueTimestamp) {
-		return Instant.now(this.clock).isBefore(Instant.ofEpochMilli(dueTimestamp));
+	private boolean maybeApplyTimingCorrection(Context context, long pollTimeout, long timeUntilDue) {
+		// Correction can only be applied to ConsumerAwareMessageListener
+		// listener instances.
+		if (context.consumer == null) {
+			return false;
+		}
+
+		boolean isInCorrectionWindow = timeUntilDue > pollTimeout && timeUntilDue <=
+				pollTimeout * POLL_TIMEOUTS_FOR_CORRECTION_WINDOW;
+
+		long correctionAmount = timeUntilDue % pollTimeout;
+		if (isInCorrectionWindow && correctionAmount > TIMING_CORRECTION_THRESHOLD) {
+			this.taskExecutor.execute(() -> doApplyTimingCorrection(context, correctionAmount));
+			return true;
+		}
+		return false;
+	}
+
+	private void doApplyTimingCorrection(Context context, long correctionAmount) {
+		try {
+			logger.debug(() -> String.format("Applying correction of %s millis at %s for TopicPartition %s",
+					correctionAmount, getCurrentMillisFromClock(), context.topicPartition));
+			this.sleeper.sleep(correctionAmount);
+			logger.debug(() -> "Waking up consumer for partition topic: " + context.topicPartition);
+			context.consumer.wakeup();
+		}
+		catch (InterruptedException e) {
+			Thread.interrupted();
+			throw new IllegalStateException("Interrupted waking up consumer while applying correction " +
+					"for TopicPartition " + context.topicPartition, e);
+		}
+		catch (Throwable e) {
+			logger.error(e, () -> "Error waking up consumer while applying correction " +
+					"for TopicPartition " + context.topicPartition);
+		}
 	}
 
 	private MessageListenerContainer getListenerContainerFromContext(Context context) {
 		return this.registry.getListenerContainer(context.listenerId);
 	}
 
 	protected void addBackoff(Context context, TopicPartition topicPartition) {
-		synchronized (this.backOffTimes) {
-			this.backOffTimes.put(topicPartition, context);
+		synchronized (this.backOffContexts) {
+			this.backOffContexts.put(topicPartition, context);
 		}
 	}
 
-	protected Context getBackoff(TopicPartition topicPartition) {
-		synchronized (this.backOffTimes) {
-			return this.backOffTimes.get(topicPartition);
+	protected Context getBackOffContext(TopicPartition topicPartition) {
+		synchronized (this.backOffContexts) {
+			return this.backOffContexts.get(topicPartition);
 		}
 	}
 
 	protected void removeBackoff(TopicPartition topicPartition) {
-		synchronized (this.backOffTimes) {
-			this.backOffTimes.remove(topicPartition);
+		synchronized (this.backOffContexts) {
+			this.backOffContexts.remove(topicPartition);
 		}
 	}
 
-	public Context createContext(long dueTimestamp, String listenerId, TopicPartition topicPartition) {
-		return new Context(dueTimestamp, listenerId, topicPartition);
+	public Context createContext(long dueTimestamp, String listenerId, TopicPartition topicPartition, Consumer<?, ?> consumer) {
```

Comment:
Can we name that `Consumer` parameter more explicit way?

Reply:
I renamed it to consumerForTimingCorrection - not sure I like it better though 😄. Let me know if you have any suggestions.